feat: add compatibility mode enable / disable button for macOS

cdrage opened the pull request:
feat: add compatibility mode enable / disable button for macOS

### What does this PR do?

* Adds a compatibility mode enable / disable button (with modal) for
  macOS users
* Runs podman-mac-helper binary that's presumingly installed via podman
  brew or pkg
* Errors out / outputs on success

In the future we will implement support for Windows and Linux
compatibility enablement.

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast explaining what is doing this PR -->

https://user-images.githubusercontent.com/6422176/225044544-63ff628b-e878-4a87-b6cb-5f15dd388e05.mov



### What issues does this PR fix or reference?

Fixes part of: https://github.com/containers/podman-desktop/issues/1447

<!-- Please include any related issue from Podman Desktop repository (or from another issue tracker).
-->

### How to test this PR?

Press the button on macOS via `yarn watch`. You should be able to
disable / enable compatibility mode successfully by following the prompt
instructions.

<!-- Please explain steps to reproduce -->

benoitf commented on package.json:
```diff
@@ -94,6 +94,7 @@
     "@docker/extension-api-client-types": "0.3.4",
     "@kubernetes/client-node": "^0.18.1",
     "@types/stream-json": "^1.7.3",
+    "@vscode/sudo-prompt": "^9.3.1",
```
it should be in package.json of the podman extension

cdrage commented:
FYI This uses `@vscode/sudo-prompt` which is a maintained library by the vscode team for running commands that require administrative access to run.

benoitf commented on extensions/podman/src/extension.ts:
```diff
@@ -451,10 +452,43 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
+    // Create a modal dialog to ask the user if they want to enable or disable compatibility mode
+    extensionApi.commands.registerCommand('podman.socketCompatibilityMode', async () => {
+      const result = await extensionApi.window.showInformationMessage(
+        `Do you want to automatically enable or disable Docker socket compatibility mode for Podman?\n\n${socketCompatibilityMode.details}`,
```
I guess as a user I would expect to have the message to disable if it's enabled or enable if it's disabled ?

Here it looks like I can enable if it's already enabled or disable if it's already disabled

cdrage commented:
@benoitf @lstocchi @slemeur @mairin 

**UI:**

Based on Slack discussion, I put this in the status bar as a place to press the button, similar to Kind as well as (future) Docker Compose binary installation. 

Let me know UI/UX suggestions. 

**Future OS':**

This will be enabled for Windows and Linux users in a separate PRs.

benoitf commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -0,0 +1,98 @@
+  };
+  return new Promise((resolve, reject) => {
+    sudo.exec(command, sudoOptions, (error, stdout, stderr) => {
+      console.log('sudo command output: ', stdout, stderr);
```
```suggestion
      console.log('sudo command output: ', stdout, stderr);
```

cdrage commented:
@benoitf Updated based on review. Will now only show enable / disable depending on the warning message regarding compatibility mode.

Also added part to restart the Podman machine after enabling / disable (required in order for socket changes).

benoitf commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -0,0 +1,100 @@
+ ***********************************************************************/
+
+import * as extensionApi from '@podman-desktop/api';
+import * as sudo from '@vscode/sudo-prompt';
```
why do we need to use the fork of VS Code instead of https://www.npmjs.com/package/sudo-prompt ?

cdrage replied:
VSCode fork is maintained and includes some bug fixes. The package you link has the last commit 3 years ago.

benoitf replied:
AFAIK vscode one only contains Microsoft stuff on top of the fork
Original one is downloaded more than 10x time than the fork

benoitf replied:
https://github.com/jorangreef/sudo-prompt/compare/master...bpasero:sudo-prompt:vscode

benoitf reviewed (commented):
code looks ok.

I was wondering if you could add some unit tests as it's a new feature.

mairin commented:
Maybe silly - is this an OS X difference? - but it seems enable should be the primary action so it should be left, and disable should be right, in the popup dialog?

cdrage commented:
Thanks! Updated it with the different package as well as swapped the buttons around so they are correct (the "enable" / disable is on the side).

slemeur commented:
In the message displayed about the compatibility mode on the dashboard, we may want to add a hint about how to enable it directly from the tool, instead of going into the troubleshooting guide ;) (not sure if you want to handle as part of this PR)
<img width="1162" alt="Capture d’écran 2023-03-14 à 19 26 09" src="https://user-images.githubusercontent.com/1636769/225102400-97acb810-c62c-4f13-9bef-fbca6a4edf49.png">

Would it be possible to use a different icon depending if the compat mode is activate or not?

cdrage commented:
> In the message displayed about the compatibility mode on the dashboard, we may want to add a hint about how to enable it directly from the tool, instead of going into the troubleshooting guide ;) (not sure if you want to handle as part of this PR) <img alt="Capture d’écran 2023-03-14 à 19 26 09" width="1162" src="https://user-images.githubusercontent.com/1636769/225102400-97acb810-c62c-4f13-9bef-fbca6a4edf49.png">
> 
> Would it be possible to use a different icon depending if the compat mode is activate or not?

Good idea! At the moment I wouldn't add it to this PR as it only affects macOS. But we should be adding something like that.

That "notification" / warning section should be changed soon.

cdrage commented:
Updated PR:
- Extra checks for binary (checks the brew folder and opt folder)
- Fixed issue with it not working on a compiled binary vs `yarn watch` (wasn't finding the binary correctly).

benoitf commented:
note sure to understand why using jest would help there ?

here is an example of mocking the API
https://github.com/containers/podman-desktop/pull/1578/files#diff-22b9841b2b26d4a6a126d06be82de9db8d68618acf39ba19ab8534502d04adafR66-R85

cdrage commented:
> note sure to understand why using jest would help there ?
> 
> here is an example of mocking the API https://github.com/containers/podman-desktop/pull/1578/files#diff-22b9841b2b26d4a6a126d06be82de9db8d68618acf39ba19ab8534502d04adafR66-R85

Ah I see, I'll add a test that uses a mock of the extension API from that referenced code.

benoitf commented on extensions/podman/src/extension.ts:
```diff
@@ -451,10 +452,56 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
+    // Create a status bar item to show the status of compatibility mode as well as
+    // create a command so when you can disable / enable compatibility mode
+    const statusBarItem = extensionApi.window.createStatusBarItem();
+    statusBarItem.text = 'Docker Socket Compatibility';
```
IMHO the text might be too long for the status bar. Tooltip is already explaining

cdrage replied:
Any suggestions? Maybe just "Socket" with a mini-docker logo? Or just "Docker Socket" (since we show the little "plug" logo anyways which looks like compatibility icon).

slemeur replied:
"Docker Compatibility"

benoitf commented on extensions/podman/src/compatibility-mode.spec.ts:
```diff
@@ -0,0 +1,95 @@
+
+  const enable = await compatibility.enable();
+  expect(enable).toBeUndefined();
+
```
should check probably that runSudoMacHelperCommand has been called as well ?

cdrage replied:
Ah, I had it here to make sure it's undefined / it did not error out. But we should be checking that the functions have been called, etc.

benoitf commented on extensions/podman/src/compatibility-mode.spec.ts:
```diff
@@ -0,0 +1,95 @@
+  expect(compatibility).toBeInstanceOf(DarwinSocketCompatibility);
+
+  const enable = await compatibility.enable();
+  expect(enable).toBeUndefined();
```
I'm not sure what this check does ? enable is a void method so not sure what are we checking on the result ?

benoitf commented on extensions/podman/src/compatibility-mode.spec.ts:
```diff
@@ -0,0 +1,95 @@
+
+  (mockDarwinSocketCompatibility.findPodmanHelper as Mock).mockResolvedValue('/opt/podman/bin/podman-mac-helper');
+  const enabled = await getSocketCompatibility().enable();
+  expect(enabled).toBeUndefined();
```
not sure here as well what we check ?

benoitf commented on extensions/podman/src/extension.ts:
```diff
@@ -451,10 +452,56 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
+    extensionContext.subscriptions.push(statusBarItem);
+
+    // Create a modal dialog to ask the user if they want to enable or disable compatibility mode
+    extensionApi.commands.registerCommand('podman.socketCompatibilityMode', async () => {
```
while testing the command was not unregistered upon save.

you need to push the result of registerCommand to `extensionContext.subscriptions.push` so it's cleaned up while deactivating the extension

benoitf commented on extensions/podman/src/extension.ts:
```diff
@@ -451,10 +452,56 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
+      const isDisguisedPodmanSocket = await isDisguisedPodman();
+
+      if (!isDisguisedPodmanSocket) {
+        const result = await extensionApi.window.showInformationMessage(
+          `Do you want to automatically enable socket compatibility mode for Podman?\n\n${socketCompatibilityMode.details}`,
+          'Enable',
+          'Close',
+        );
```
It looks like I have this dialog

![image](https://user-images.githubusercontent.com/436777/225746494-5ec7d20f-a1cf-488e-9a6d-f062cb46cc59.png)

benoitf replied:
while if I replace `Close` by `Cancel` I have:
![image](https://user-images.githubusercontent.com/436777/225746765-c0e9b428-0216-4a1a-bd50-1b7fbd84f4dc.png)


so I have a primary action/button enable which looks nicer to me . WDYT ?

benoitf commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -0,0 +1,112 @@
+    };
+    return new Promise((resolve, reject) => {
+      sudo.exec(command, sudoOptions, (error, stdout, stderr) => {
+        console.log('sudo command output: ', stdout, stderr);
```
```suggestion

```

benoitf commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -0,0 +1,112 @@
+    try {
+      await this.runSudoMacHelperCommand(fullCommand);
+      extensionApi.window.showInformationMessage(`Docker socket compatibility mode for Podman has been ${description}.
+      Restart your Podman machine to apply the changes.`);
```
I might expect to have a message prompting me: click to Restart now/Later and the extension restart for me the machine for me (could be a separate/later improvement) but I think UX would be better

benoitf reviewed (commented):
I have a UX issue with the status bar element.

When I hover the element it says: Enable or disable the docker socket compatibility

but it knows if it's enabled and working or disabled so I would expect to see the state in the tooltip/hover
is that the icon should be different as well when mode is enabled vs when mode is disabled.

also as we're just suggesting that we need to restart the machine and not restarting it, we can click again on the icon that will prompt: would you like to enable

but if we do it it will result in
![image](https://user-images.githubusercontent.com/436777/225748263-8c4a08ef-14d5-4686-9cb6-f6961d54bc06.png)

so by checking the file `/Library/LaunchDaemons/com.github.containers.podman.helper-<username>.plist` we know if it's already enabled and that we don't have restarted the machine as well so we can be smarter in the report.

benoitf commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -0,0 +1,121 @@
+
+import * as extensionApi from '@podman-desktop/api';
+import * as sudo from 'sudo-prompt';
+import * as fs from 'fs';
```
```suggestion
import * as fs from 'node:fs';
```

benoitf replied:
then we're sure nobody will fetch https://www.npmjs.com/package/fs :-)

cdrage replied:
thank you! forgot about that :)

benoitf commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -0,0 +1,121 @@
+import * as extensionApi from '@podman-desktop/api';
+import * as sudo from 'sudo-prompt';
+import * as fs from 'fs';
+import * as os from 'os';
```
```suggestion
import * as os from 'node:os';
```

cdrage commented:
Updated PR:
* Only show enable / disable if the plist file exists or not
* Show primary / blue button now (didn't realize if you put just "Cancel" it'll make the other blue!)
* Opened up issue: https://github.com/containers/podman-desktop/issues/1733 so I can work on a button to restart gracefully / nicely.

benoitf commented:
@cdrage is it possible to update the tooltip to reflect what will trigger the click on the icon/status bar element ?

benoitf reviewed (approved):
was able to enable/disable the compatibility mode of Podman

would be nice if tooltip would reflect what the action will done or the current mode
like : 'Compatibility is enabled. Click to disable' or 'Compatibility is disabled: Click to enable'

cdrage commented:
> was able to enable/disable the compatibility mode of Podman
> 
> would be nice if tooltip would reflect what the action will done or the current mode like : 'Compatibility is enabled. Click to disable' or 'Compatibility is disabled: Click to enable'
> 
> and #1733 will help UX as we see that mode is enabled or disabled but it's not reflected without restarting the machine

Added! See latest commit :)

cdrage commented:
> was able to enable/disable the compatibility mode of Podman
> 
> would be nice if tooltip would reflect what the action will done or the current mode like : 'Compatibility is enabled. Click to disable' or 'Compatibility is disabled: Click to enable'
> 
> and #1733 will help UX as we see that mode is enabled or disabled but it's not reflected without restarting the machine

Yeah, for #1733 I was getting minor issues when restarting the podman machine via the button (sometimes stopping would time out, but it wouldn't start after...) could be my podman machine not working correctly (have to rm and recreate), but wasn't too sure. 

I agree, let's get #1733 in as a future improvement.

benoitf reviewed (approved):
approving again
